pkg/proxy/backup.go

```diff
@@ -0,0 +1,407 @@
+package proxy
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+
+	rpc "github.com/longhorn/longhorn-instance-manager/pkg/imrpc"
+
+	eclient "github.com/longhorn/longhorn-engine/pkg/controller/client"
+	rclient "github.com/longhorn/longhorn-engine/pkg/replica/client"
+	esync "github.com/longhorn/longhorn-engine/pkg/sync"
+	etypes "github.com/longhorn/longhorn-engine/pkg/types"
+	eutil "github.com/longhorn/longhorn-engine/pkg/util"
+	eptypes "github.com/longhorn/longhorn-engine/proto/ptypes"
+
+	"github.com/longhorn/backupstore"
+	bsutil "github.com/longhorn/backupstore/util"
+)
+
+func (p *Proxy) SnapshotBackup(ctx context.Context, req *rpc.EngineSnapshotBackupRequest) (resp *rpc.EngineSnapshotBackupProxyResponse, err error) {
+	log := logrus.WithFields(logrus.Fields{"serviceURL": req.ProxyEngineRequest.Address})
+	log.Debugf("Backing up snapshots %v to backup %v", req.SnapshotName, req.BackupName)
+
+	for _, env := range req.Envs {
+		part := strings.SplitN(env, "=", 2)
+		if len(part) < 2 {
+			continue
+		}
+
+		if err := os.Setenv(part[0], part[1]); err != nil {
+			return nil, err
+		}
+	}
+
+	credential, err := eutil.GetBackupCredential(req.BackupTarget)
+	if err != nil {
+		return nil, err
+	}
+
+	labels := []string{}
+	for k, v := range req.Labels {
+		labels = append(labels, fmt.Sprintf("%s=%s", k, v))
+	}
+
+	task, err := esync.NewTask(ctx, req.ProxyEngineRequest.Address)
+	if err != nil {
+		return nil, err
+	}
+
+	recv, err := task.CreateBackup(
+		req.BackupName,
+		req.SnapshotName,
+		req.BackupTarget,
+		req.BackingImageName,
+		req.BackingImageChecksum,
+		labels,
+		credential,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &rpc.EngineSnapshotBackupProxyResponse{
+		BackupId:      recv.BackupID,
+		Replica:       recv.ReplicaAddress,
+		IsIncremental: recv.IsIncremental,
+	}, nil
+}
+
+func (p *Proxy) SnapshotBackupStatus(ctx context.Context, req *rpc.EngineSnapshotBackupStatusRequest) (resp *rpc.EngineSnapshotBackupStatusProxyResponse, err error) {
+	log := logrus.WithFields(logrus.Fields{"serviceURL": req.ProxyEngineRequest.Address})
+	log.Debugf("Getting %v backup status from replica %v", req.BackupName, req.ReplicaAddress)
+
+	c, err := eclient.NewControllerClient(req.ProxyEngineRequest.Address)
+	if err != nil {
+		return nil, err
+	}
+	defer c.Close()
+
+	replicas, err := p.ReplicaList(ctx, req.ProxyEngineRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	status := &esync.BackupStatusInfo{}
+	replicaAddress := req.ReplicaAddress
```

Comment:
NIT: 
```
var  status *esync.BackupStatusInfo

......

if status == nil {
	return nil, errors.Errorf("failed to find a replica with backup %s", req.BackupName)
}
```

Reply:
@c3y1huang  you can follow up with this later. Thanks.